### ISSUE TITLE
pytest-runner should be in test-requirements

### ISSUE THREAD

**Issue by jaraco** [jaraco/keyring#105]
It doesn't seem to be used to install or do anything after install (except for running tests).

---
- Bitbucket: https://bitbucket.org/kang/python-keyring-lib/issue/105
- Originally reported by: Matthew Thode
- Originally created at: 2013-07-10T22:02:59.446

**Comment by jaraco** [comment]
You're right. `pytest-runner` doesn't do anything during install. In fact, it doesn't do anything during any built-in setuptools or distutils command. Instead, it provides its own setup command 'ptr'.

pytest-runner cannot be in test_requirements because it must be present before test_requirements is honored. It must necessarily already be installed in order to supply the setuptools plugin to supply the 'ptr' command (which in turn honors test_requirements and runs the tests). It's a shame that it must also be present for installing the package even though it's not invoked.

Perhaps something else can be done. Maybe it would make sense to only add the setup requirement when 'ptr' appears in sys.argv, although that feels hacky. What is the motivation for changing it?

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
We (Gentoo) mainly want to reduce the dependencies on it (since pytest-runner has it's own deps).  It seems like we need a test-dependencies-dependencies thing :P

---

Original comment by: Matthew Thode

**Comment by jaraco** [comment]
Thinking on it more, I think that maybe pytest-runner should just be a dep of one of the dependencies of test_requirements.

Which package requires it?  That package should get a bug against it (for not listing it in it's requirements).

---

Original comment by: Matthew Thode

**Comment by jaraco** [comment]
Hi promethianfire. I don't believe there's any way to move the dependency. keyring requires pytest-runner _before_ tests_requires is even invoked, so for the test dependencies to declare a dependency on pytest runner would not help.

The point of a 'setup_requires' dependency is to signal to setuptools that this requirement is necessary when setup is run (for build/test/install commands or to provide additional commands). The only package that could possibly require or bundle pytest-runner is setuptools itself, but that's not going to happen (because that approach isn't scalable).

But perhaps there's something that can be done to ease the burden on Gentoo. Since the package is only required when running setup, it should not be required at run time (similar to tests_require) nor if keyring is installed by a system package manager (such as a Windows installer or Linux system package). This is generally true for all 'setup_requires' dependencies.

This fact gives Gentoo the opportunity to run setup.py once to generate the package (pulling in pytest-runner during that step) but then bundle the build and disregard pytest-runner thereafter.

Can you clarify when the dependency on pytest runner is problemmatic? Is it during the preparation of Gentoo packages or later when the users install keyring?

I'm still exploring this landscape myself and trying to understand the complications imposed by 'setup_requires' declarations. I use the declaration in many packages throughout my work every day, and it works well, but I work in an environment which has access to the packages it needs and installs them with pip/setuptools directly. I expect I might even want to add to the setup_requires dependencies. If doing so is going to complicate the system package maintainers, I'd like to know what those complications are or devise a way to minimize that effect.

Can you elaborate on the challenges this dependency creates?

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
test_requires=['package_that_needs_pytest-runner', package_that_does_not_need_pytest-runner']

I'm saying that package_that_needs_pytest-runner should have pytest-runner in it's setup_reuqires, sense it's required.

There's no place that it's problematic, but it does require us to install things we don't use.  If I move the dependency over to the testing side on gentoo then I can't guarantee that it's installed before the other stuff.

the *_requires dependencies should only be what the program itself needs, not what another dependency needs.  Where is pytest-runner called within 'python-keyring-lib'  or it's tests?

---

Original comment by: Matthew Thode

**Comment by jaraco** [comment]
I think I may have a solution, does pytest-runner need to be installed before the testing packages are installed or does it only need to be installed before the tests are run?

---

Original comment by: Matthew Thode

**Comment by jaraco** [comment]
None of the packages in tests_require have any dependency on pytest-runner. The thing that requires pytest-runner is the person or process that runs 'setup.py ptr', such as the Travis-CI script (.travis.yml) does. So really, it's keyring's recommended way of invoking tests that requires pytest-runner.

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
Here's [where pytest-runner is used](/kang/python-keyring-lib/src/59a8b11c2a6828fb2ed10370acdd08ae12c0af6f/.travis.yml?at=default#cl-12).

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
so I can ignore that, it should be something within travis/jenkins that has it as a prereq, not the package itself :D

maybe an additional function if that script should install pytest-runner, that seems like the proper place to me.

---

Original comment by: Matthew Thode

**Comment by jaraco** [comment]
ok, since it's only needed for tests being run (not test depends being installed), I've moved it to test depends on our side.  I think we can close this now, thanks for your patience :D

---

Original comment by: Matthew Thode

**Comment by jaraco** [comment]
In Debian we also have a [patch](http://anonscm.debian.org/viewvc/python-modules/packages/python-keyring/trunk/debian/patches/no-pytest-runner.patch?view=markup) that removes `pytest-runner` from `setup_requires`. It will be nice if it was only added when `ptr` appears in `sys.argv`, as Jason suggested.

---

Original comment by: Dmitry Shachnev

**Comment by jaraco** [comment]
Issue #106 was marked as a duplicate of this issue.

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
In #106, @dhellmann suggests (and sends a [corresponding pull request](/kang/python-keyring-lib/pull-request/34)) that the requirement simply be removed.

However, it's not so simple to simply remove the requirement because it's inconvenient. The purpose of the requirement, as described above, is to enable a person or system to be able to invoke the tests without having to have anything special in his environment (except Python and setuptools).

The setup_requires declaration takes advantage of the setuptools machinery to ensure that packages necessary to invoke the full operation of the setup script are available. It does this for any environment and without any prerequisite steps. It's these advantages that are particularly valuable because it factors out those prerequisite steps into a single, declarative indicator. If this declaration were removed, it would require (a) the documentation to be updated to indicate how to set up an environment to test, (b) the travis CI tests (and any other automated test scripts) to be updated to fulfill this need, and (c) testers would be responsible for creating a suitable environment in which to meet these prerequisites, running the tests, and then resetting the environment to the state prior to the tests. The whole testing process becomes a lot more complicated without support for downloading the requirements in advance.

This need would be obviated if there were a way to invoke py.test from setuptools without the 'setup_requires' declaration (like there is for nose and unittest), but I don't believe there is.

We can work around the issue by suppressing the presence of 'pytest-runner' in the setup_requires unless 'ptr' is in the command-line. It's a bit of a hack, and inelegant, but far superior to simply removing it.

prometheanfire indicated that the issue 'is not problemmatic', so I considered the issue low priority. As @dhellmann re-raises the issue with respect to OpenStack, I'd like to re-pose the question - what is the underlying issue that OpenStack is unable to acquire pytest-runner when installing keyring?

The other issue to consider here is I'd like to add another 'setup_requires' requirement to keyring to enable mercurial repository support and version tagging (via [hgtools](/jaraco/hgtools)). If the presence of a pytest-runner requirement is indicative of a general problem with setup_requires dependencies, then I'd like to explore solving that problem in general, rather than simply suppressing it for this particular case.

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
Have you considered using tox to wrap the test run? That would let you move pytest-runner from setup.py to the dependencies installed via tox, so it would be present when tox runs "python setup.py ptr".

---

Original comment by: Doug Hellmann

**Comment by jaraco** [comment]
There are 2 problems as it stands now.
1. We don't want pytest-runner installed because it is not one of our dependencies. Production systems running OpenStack don't need it.
2. There is some sort of problem with pip figuring out the second or third level dependency on pytest-runner that breaks when running our tests. We're working on fixing that, but even when it works we will have (1).

---

Original comment by: Doug Hellmann

**Comment by jaraco** [comment]
I would consider using tox, but I don't see how that helps the situation. There's no way to invoke tox without installing it first, which means all of the same bootstrapping issues still arise (where is tox installed, how is it invoked, how is it removed). I'm trying to avoid saddling the tester with these details. I believe the Travis-CI story would be a good one if it comes with tox pre-installed, so that's part of the story.

I don't think (1) is a problem. One nice detail about the 'setup-requires' dependencies is they're not installed permanently, but only ephemerally (for the invocation of setup.py itself). Thereafter, the egg may still be present on the file system, but pytest-runner is not and does not need to be installed. And indeed, it isn't:

```
vagrant@precise64:~$ tkey/bin/pip install keyring
Downloading/unpacking keyring
  Downloading keyring-1.6.zip (82kB): 82kB downloaded
  Running setup.py egg_info for package keyring
    zip_safe flag not set; analyzing archive contents...

    Installed /tmp/easy_install-VHtGTs/pytest-runner-2.0/hgtools-3.0.2-py2.7.egg

    Installed /home/vagrant/tkey/build/keyring/pytest_runner-2.0-py2.7.egg

    warning: no previously-included files found matching '.hg/last-message.txt'
Installing collected packages: keyring
  Running setup.py install for keyring

    warning: no previously-included files found matching '.hg/last-message.txt'
    Installing keyring script to /home/vagrant/tkey/bin
Successfully installed keyring
Cleaning up...
vagrant@precise64:~$ tkey/bin/pip freeze
argparse==1.2.1
keyring==1.6
wsgiref==0.1.2
```

Therefore, if (2) can be resolved, then there are no further issues?

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
@jaraco actually, there is a tox bootstrap script that would help this situation. If you have the bootstrap script then you don't need to declare any dependencies at all since the script takes care of pulling in all the necessary things for testing.

Tox Bootstrap: https://bitbucket.org/hpk42/tox/src/fb612de381a79f8d7bff9cf2e0bc8fab42237efd/toxbootstrap.py

Even if there are ephemeral installations of packages, they still represent a problem if the system installing the package does not require them.

---

Original comment by: Alfredo Deza

**Comment by jaraco** [comment]
Fix #105: Only include pytest-runner in 'setup requirements' when ptr invocation is indicated in the command-line.

→ <<cset 29116566d46e>>

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
Thanks, Jason!

---

Original comment by: Doug Hellmann

**Comment by jaraco** [comment]
Issue #106 was marked as a duplicate of this issue.

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
As it seems clear that there's a certain annoyance factor with requiring pytest runner at install time (even though it's not installed), I've added a small check to exclude the requirement if it appears it's not needed. It's been released as 1.6.1. Please let me know if that doesn't make everyone's life easier.

---

Original comment by: Jason R. Coombs
